CSV Validation Research

TeaSmith7 opened the pull request:
The work involves analysis of various CSV libraries across different programming languages, including Kotlin, Go's CSV package, Rust's CSV package, and Microsoft.VisualBasic.FileIO.TextFieldParser for .NET. The main focus of the work involved the creation of functions to validate CSV files containing both 100 and 10,000 rows, ensuring compliance with the RFC-4180 standard. Additionally, unit tests were developed specifically to evaluate RFC-4180 compliance for both the Go and OpenCSV library. Performance metrics are included in the README files to provide insights into the efficiency and effectiveness of each library's CSV parsing capabilities.

TeaSmith7 committed:
This code reads the csv file from data directory processes
CSV data and performs profiling. The main function reads data from a
CSV file, iterates over its records, and performs no operation on each
field. Additionally, CPU and memory profiling are implemented using the
go's built in 'pprof' package to analyze the performance of the program for CPU and
Memory Usage.

TeaSmith7 committed:
deleted files created by pprof

TeaSmith7 committed:
This code reads harcoded csv file from the data directory
 - read_csv function handles reading each row and prints records
 - Added error handling for row level errors
 = prints elapsed time for CSV reading operations(# executions can be
 changed with outer for loop)

TeaSmith7 committed:
Analyze CSV parsing  performance with multiple libraries in Kotlin:
The CSV parsing libraries analyzed include:
- kotlin-csv
- OpenCSV
- Univocity parsers
- Apache Commons CSV
- Deephaven CSV

Each library is implemented in its own method within the CSVValidator
class:
- validateWithKotlinCSV
- validateWithOpenCSV
- validateWithApacheCommons
- univocityCSVParser
- deepHavenCSVParser

TeaSmith7 committed:
updates to Main.kt

TeaSmith7 committed:
updated validation results table for rust

TeaSmith7 committed:
Implemented CSV Reader in C# within .NET framework
function that reads harcoded filename stored under data folder
iterates through and stores row in data List
errors found are stored in errors List
custom error handling can also be implemented

TeaSmith7 committed:
moved line that reads csv data from the path outside the loop
updated the name of the test csv file to th
e correct number of rows and reflected the change to the table in the
readme.

TeaSmith7 committed:
added tests for CR, LF, CRLF
Renamed file to correct number of rows and updated README.md accordingly

TeaSmith7 committed:
Added unit test for rfc-4180 compliance
Each record should contain the same number of fields throughout the file
Within the header and within each record, there may be one or more
fields, separated by commas
csv file without header
Line breaks on the last record

TeaSmith7 committed:
added unit test for Spaces are considered part of a field and should not
be ignored.
deleted previous file with wrong number of rows

TeaSmith7 committed:
added tests for TestLastFieldInRecordFollowedByComma() and
TestQuotesInFieldNotEnclosedWithDoubleQuotes

TeaSmith7 committed:
added tests for
TestDoubleQuotesInsideQuotedField
TestFieldsWithLineBreaksInsideQuotes
TestFieldsWithCommasInsideQuotes

TeaSmith7 committed:
deleted kotlin folder

TeaSmith7 committed:
moved kotlin folder under csv-validation-research

TeaSmith7 committed:
added placeholder unit tests for rfc-4180 compliance for
openCSV
ApacheCommons
kotlin-csv
UnivocityParser

TeaSmith7 committed:
started creating unit tests using OpenCSV library for RFC-4180
compliance

TeaSmith7 committed:
added more tests and updated table in README.md

TeaSmith7 committed:
completed remaining tests for openCSV library against RFC-4180 standard
updated the README.md with the results
deleted some empty classes

TeaSmith7 committed:
deleted last two files remaining under kotlin folder

TeaSmith7 committed:
updated existing csv libraries to iterate through the row and get field
removed print statements
added functions for Super CSV, Jackson CSV, and Fast CSV.

TeaSmith7 committed:
added benchmark test that reads csv file and reports allocations

TeaSmith7 committed:
updated the readme with steps to benchmark

TeaSmith7 committed:
added papaparse csv file library
Changes:
- reads csv file from data folder, for each row gets the field
- added benchmarking to record cpu usage and memory usage

TeaSmith7 committed:
updated each csv library function with:
data structure that holds list of row objects

TeaSmith7 committed:
creating row object and appending to rows, which is declares as slice
that holds Row structs.

TeaSmith7 committed:
updated README with performance metrics and added results for
Fast CSV
Super CSV
Jackson CSV

TeaSmith7 committed:
updated memory allocation for FastCSV.

TeaSmith7 committed:
corrected readme with steps to run benchmarking

erik1066 committed:
Fixed broken Go benchmark

TeaSmith7 committed:
corrected getting memory metrics.
getting entire metrics that includes heapUsed

TeaSmith7 committed:
removed file committed by error

TeaSmith7 committed:
removed python folder committed by error

TeaSmith7 committed:
corrected benchmarking code:
Added 'complete' callback that is triggered once parsing is finished,
to make sure benchmarking waits until all parsing is done before
completing.
Also returning memory metrics and converting from Bytes to MB

erik1066 commented on csv-validation-research/dotnet/README.md:
```diff
@@ -0,0 +1,18 @@
+| 100 |15100|||file-with-headers-10000-rows.csv|
+| 5000|847707|||file-with-headers-10000-rows.csv|
+
+##How to run the code
```
Doesn't appear as correct header level 2.

TeaSmith7 committed:
Update README.md

corrected the header formatting issue and updated the readme.

TeaSmith7 committed:
addresses sonar cloud warnings

erik1066 committed:
Update README.md for .NET research

Updated the README.md file that's specific to the .NET research. The language mentioned Rust instead of C# in one place. I also added some helpful links and the specific version of .NET needed to run this code.

erik1066 committed:
Update CsvReader.cs

Made minor code spacing improvements to align with C# language guidelines.

erik1066 committed:
Update README.md

Minor updates and added some URLs.

erik1066 reviewed (approved):
Approved
